src/cargo/core/compiler/mod.rs

```diff
@@ -1976,6 +2008,37 @@ fn on_stderr_line_inner(
         return Ok(false);
     }
 
+    // Returns `true` if the diagnostic was modified.
+    let extra_diag_message = |diag: &mut String| -> bool {
+        static PRIV_DEP_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+            Regex::new("from private dependency '([A-Za-z0-9-_]+)' in public interface").unwrap()
+        });
+        if let Some(crate_name) = PRIV_DEP_REGEX.captures(diag).and_then(|m| m.get(1)) {
+            let crate_name = crate_name.as_str();
+            if let Some(span) = find_crate_span(crate_name, &manifest.spans) {
+                let rel_path = pathdiff::diff_paths(&manifest.path, &manifest.cwd)
+                    .unwrap_or_else(|| manifest.path.clone())
+                    .display()
+                    .to_string();
+                let report = [Group::with_title(
+                    Level::NOTE
+                        .secondary_title(format!("dependency `{crate_name}` declared here",)),
+                )
+                .element(
+                    Snippet::source(&manifest.contents)
+                        .path(rel_path)
+                        .annotation(AnnotationKind::Context.span(span)),
+                )];
+
+                let rendered = Renderer::styled().render(&report);
+                diag.push('\n');
+                diag.push_str(&rendered);
```

Comment:
This needs to be swapped around, as there shouldn't be a newline between the original message and this one
```suggestion
                diag.push_str(&rendered);
                diag.push('\n');
```

Reply:
I tried swapping them around. As-is, the PR gives

```text
src/lib.rs:2:1: warning: type `FromPriv` from private dependency 'priv_dep' in public interface
note: dependency `priv_dep` declared here
 --> ../../tmp/test-project/Cargo.toml:9:65
  |
9 | priv_dep = { version = "0.1.0", path = "../priv_dep", package = "priv_dep" }
  |
warning: `foo` (lib) generated 1 warning
```

After swapping, I get:
```text
src/lib.rs:2:1: warning: type `FromPriv` from private dependency 'priv_dep' in public interfacenote: dependency `priv_dep` declared here
 --> ../../tmp/test-project/Cargo.toml:9:65
  |
9 | priv_dep = { version = "0.1.0", path = "../priv_dep", package = "priv_dep" }
  |                                                                 ----------

warning: `foo` (lib) generated 1 warning
```
which looks worse to me. I mean, at the very least it needs a space between "interface" and "note". But even with the space, that line is quite long...